CopyFiles alias throws exception if empty enumeration is passed

### What You Are Seeing?
The [CopyFiles](http://cakebuild.net/api/Cake.Common.IO/FileAliases/8B962704) alias currently throws an "Sequence contains no matching element" exception if an empty enumeration is passed to copy.

### What is Expected?
Either the alias should not throw (since there is nothing to copy) or an `ArgumentException` should be thrown.

### What version of Cake are you using?
0.20.0

### Are you running on a 32 or 64 bit system?
64-Bit

### What environment are you running on?  Windows? Linux? Mac?
Windows

@pascalberger @augustoproiete 
Please provide a way to reproduce this issue.

I tried with below code and got this error in 1.1.0
```

CopyFiles(
    new[]
    {
        //new FilePath("filesource/innerfolder/File1.txt"),
        //new FilePath("filesource/innerfolder/File2.txt")
    },
    "Destination", 
    preserveFolderStructure: true);
```

Errror

```
>build.ps1

C:/CodeContribute/exampleapp/build.cake(8,5): error CS0826: No best type found for implicitly-typed array
C:/CodeContribute/exampleapp/build.cake(8,5): error CS1503: Argument 1: cannot convert from '?[]' to 'Cake.Core.IO.GlobPattern'
Error: Error(s) occurred when compiling build script:
C:/CodeContribute/exampleapp/build.cake(8,5): error CS0826: No best type found for implicitly-typed array
C:/CodeContribute/exampleapp/build.cake(8,5): error CS1503: Argument 1: cannot convert from '?[]' to 'Cake.Core.IO.GlobPattern'
```

@augustoproiete 

Tried above command and I still getting same error.
Cake versions: 1.1.0
.NET CORE 5.0

```
>build.ps1

C:/CodeContribute/exampleapp/build.cake(8,5): error CS0826: No best type found for implicitly-typed array
C:/CodeContribute/exampleapp/build.cake(8,5): error CS1503: Argument 1: cannot convert from '?[]' to 'Cake.Core.IO.GlobPattern'
Error: Error(s) occurred when compiling build script:
C:/CodeContribute/exampleapp/build.cake(8,5): error CS0826: No best type found for implicitly-typed array
C:/CodeContribute/exampleapp/build.cake(8,5): error CS1503: Argument 1: cannot convert from '?[]' to 'Cake.Core.IO.GlobPattern'
```

Also, I tried with below config and got different error 
cake version:0.20.0
.NET CORE 2.1

```
error NU1102: Unable to find package cake.tool with version (= 0.20.0)
error NU1102:   - Found 21 version(s) in nuget.org [ Nearest version: 0.30.0 ]
error NU1102:   - Found 0 version(s) in Microsoft Visual Studio Offline Packages

Package "cake.tool" failed to restore, due to Microsoft.DotNet.ToolPackage.ToolPackageException: The tool package could not be restored.
   at Microsoft.DotNet.Tools.Tool.Install.ProjectRestorer.Restore(FilePath project, PackageLocation packageLocation, String verbosity)
   at Microsoft.DotNet.ToolPackage.ToolPackageInstaller.InstallPackageToExternalManagedLocation(PackageLocation packageLocation, PackageId packageId, VersionRange versionRange, String targetFramework, String verbosity)
   at Microsoft.DotNet.Tools.Tool.Restore.ToolRestoreCommand.InstallPackages(ToolManifestPackage package, Nullable`1 configFile)

Restore failed.
```

So learning from error, I changed config to below and got another error
cake version:0.30.0
.NET CORE 2.1

```
Tool 'cake.tool' (version '0.30.0') was restored. Available commands: dotnet-cake

Restore was successful.
C:/CodeContribute/exampleapp/build.cake(8,5): error CS0826: No best type found for implicitly-typed array
C:/CodeContribute/exampleapp/build.cake(8,5): error CS1503: Argument 1: cannot convert from '?[]' to 'string'
Error: Error(s) occurred when compiling build script:
C:/CodeContribute/exampleapp/build.cake(8,5): error CS0826: No best type found for implicitly-typed array
C:/CodeContribute/exampleapp/build.cake(8,5): error CS1503: Argument 1: cannot convert from '?[]' to 'string'
```

So I am interested to know to 
1. for which version of cake I should fix this issue?
2. Also, please confirm appropriate behavior i.e. to throw (since there is nothing to copy) or an ArgumentException should be thrown

@augustoproiete 
The problem here is the argument is not getting properly validated and reaching the Alias methods where we do the actual filecopy operation.

I tried my best to find the code which validates it and I got nothing so far. 
Any ideas on where and how in code we validate the arguments provided in build.cake script will be helpful

To demonstrate the validation part, I have used the below command and its breaking with error:
`error CS1503: Argument 1: cannot convert from 'Cake.Core.IO.DirectoryPath[]' to 'Cake.Core.IO.GlobPattern'`

```
CopyFiles(
    new DirectoryPath[0],
    MakeAbsolute(new DirectoryPath("Destination")),
    preserveFolderStructure: true);
}
```

This is not even reaching the Alias methods which is good thing and some validation is breaking code.
I recommend to extend this validation for the issue at hand as well. IMHO, this will make good code clearly separating script validation from script run.

<details>
<summary>@coder2213 If you run the Cake script in diagnostic mode, you'll be able to see the stack trace... (<strong>click here</strong> to expand)</summary>
<p>

```
dotnet cake test-copy-files-gh-1663.cake --verbosity diagnostic

Error: System.AggregateException: One or more errors occurred. (Sequence contains no matching element)
 ---> System.InvalidOperationException: Sequence contains no matching element
   at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
   at Cake.Common.IO.FileCopier.CopyFiles(ICakeContext context, IEnumerable`1 filePaths, DirectoryPath targetDirectoryPath, Boolean preserverFolderStructure) in C:\projects\cake\src\Cake.Common\IO\FileCopier.cs:line 112
   at Cake.Common.IO.FileAliases.CopyFiles(ICakeContext context, IEnumerable`1 filePaths, DirectoryPath targetDirectoryPath, Boolean preserveFolderStructure) in C:\projects\cake\src\Cake.Common\IO\FileAliases.cs:line 191
   at Submission#0.CopyFiles(IEnumerable`1 filePaths, DirectoryPath targetDirectoryPath, Boolean preserveFolderStructure)
   at Submission#0.<<Initialize>>d__0.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.RunSubmissionsAsync[TResult](ImmutableArray`1 precedingExecutors, Func`2 currentExecutor, StrongBox`1 exceptionHolderOpt, Func`2 catchExceptionOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Scripting.Script`1.RunSubmissionsAsync(ScriptExecutionState executionState, ImmutableArray`1 precedingExecutors, Func`2 currentExecutor, Func`2 catchExceptionOpt, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Cake.Infrastructure.Scripting.RoslynScriptSession.Execute(Script script) in C:\projects\cake\src\Cake\Infrastructure\Scripting\RoslynScriptSession.cs:line 142
   at Cake.Core.Scripting.ScriptRunner.Run(IScriptHost host, FilePath scriptPath) in C:\projects\cake\src\Cake.Core\Scripting\ScriptRunner.cs:line 172
   at Cake.Features.Building.BuildFeature.RunCore(IRemainingArguments arguments, BuildFeatureSettings settings) in C:\projects\cake\src\Cake\Features\Building\BuildFeature.cs:line 99
   at Cake.Features.Building.BuildFeature.Run(IRemainingArguments arguments, BuildFeatureSettings settings) in C:\projects\cake\src\Cake\Features\Building\BuildFeature.cs:line 49
   at Cake.Commands.DefaultCommand.Execute(CommandContext context, DefaultCommandSettings settings) in C:\projects\cake\src\Cake\Commands\DefaultCommand.cs:line 72
```
</p>
</details>

And that shows you where the exception is being thrown: [Cake.Common/IO/FileCopier.cs:line 112](https://github.com/cake-build/cake/blob/a1893a2bba26f49e089d66beeb78421cedf6126d/src/Cake.Common/IO/FileCopier.cs#L112)

---

No additional validation is required because the `error CS1503` you get is a compilation error that is already the validation. [The method that copies files](https://github.com/cake-build/cake/blob/a1893a2bba26f49e089d66beeb78421cedf6126d/src/Cake.Common/IO/FileCopier.cs#L63) expects an object that implements `IEnumerable<FilePath>`, so if you give it an object that doesn't implement `IEnumerable<FilePath>` you get the compilation error as expected.

```csharp
public static void CopyFiles(ICakeContext context,
IEnumerable<FilePath> filePaths,
DirectoryPath targetDirectoryPath,
bool preserverFolderStructure)
```

--- a/src/Cake.Common/IO/FileCopier.cs
+++ b/src/Cake.Common/IO/FileCopier.cs
@@ -106,42 +106,45 @@ namespace Cake.Common.IO
                 throw new DirectoryNotFoundException(message);
             }
 
-            if (preserverFolderStructure)
+            if (filePaths.Any())
             {
-                var commonPath = string.Empty;
-                var separatedPath = filePaths
-                    .First(str => str.ToString().Length == filePaths.Max(st2 => st2.ToString().Length)).ToString()
-                    .Split(new string[] { "/" }, StringSplitOptions.RemoveEmptyEntries)
-                    .ToList();
-
-                foreach (string pathSegment in separatedPath)
+                if (preserverFolderStructure)
                 {
-                    if (commonPath.Length == 0 && filePaths.All(str => str.ToString().StartsWith(pathSegment)))
+                    var commonPath = string.Empty;
+                    var separatedPath = filePaths
+                        .First(str => str.ToString().Length == filePaths.Max(st2 => st2.ToString().Length)).ToString()
+                        .Split(new string[] { "/" }, StringSplitOptions.RemoveEmptyEntries)
+                        .ToList();
+
+                    foreach (string pathSegment in separatedPath)
                     {
-                        commonPath = pathSegment;
+                        if (commonPath.Length == 0 && filePaths.All(str => str.ToString().StartsWith(pathSegment)))
+                        {
+                            commonPath = pathSegment;
+                        }
+                        else if (filePaths.All(str => str.ToString().StartsWith(commonPath + "/" + pathSegment)))
+                        {
+                            commonPath += "/" + pathSegment;
+                        }
+                        else
+                        {
+                            break;
+                        }
                     }
-                    else if (filePaths.All(str => str.ToString().StartsWith(commonPath + "/" + pathSegment)))
+
+                    // Iterate all files and copy them.
+                    foreach (var filePath in filePaths)
                     {
-                        commonPath += "/" + pathSegment;
-                    }
-                    else
-                    {
-                        break;
+                        CopyFileCore(context, filePath, absoluteTargetDirectoryPath.GetFilePath(filePath), context.DirectoryExists(commonPath) ? commonPath : null);
                     }
                 }
-
-                // Iterate all files and copy them.
-                foreach (var filePath in filePaths)
+                else
                 {
-                    CopyFileCore(context, filePath, absoluteTargetDirectoryPath.GetFilePath(filePath), context.DirectoryExists(commonPath) ? commonPath : null);
-                }
-            }
-            else
-            {
-                // Iterate all files and copy them.
-                foreach (var filePath in filePaths)
-                {
-                    CopyFileCore(context, filePath, absoluteTargetDirectoryPath.GetFilePath(filePath), null);
+                    // Iterate all files and copy them.
+                    foreach (var filePath in filePaths)
+                    {
+                        CopyFileCore(context, filePath, absoluteTargetDirectoryPath.GetFilePath(filePath), null);
+                    }
                 }
             }
         }

--- a/src/Cake.Common/IO/FileCopier.cs
+++ b/src/Cake.Common/IO/FileCopier.cs
@@ -106,45 +106,48 @@ namespace Cake.Common.IO
                 throw new DirectoryNotFoundException(message);
             }
 
-            if (filePaths.Any())
+            if (preserverFolderStructure)
             {
-                if (preserverFolderStructure)
+                var commonPath = string.Empty;
+                var separatedPath = filePaths
+                    .First(str => str.ToString().Length == filePaths.Max(st2 => st2.ToString().Length)).ToString()
+                    .Split(new string[] { "/" }, StringSplitOptions.RemoveEmptyEntries)
+                    .ToList();
+
+                foreach (string pathSegment in separatedPath)
                 {
-                    var commonPath = string.Empty;
-                    var separatedPath = filePaths
-                        .First(str => str.ToString().Length == filePaths.Max(st2 => st2.ToString().Length)).ToString()
-                        .Split(new string[] { "/" }, StringSplitOptions.RemoveEmptyEntries)
-                        .ToList();
-
-                    foreach (string pathSegment in separatedPath)
+                    if (commonPath.Length == 0 && filePaths.All(str => str.ToString().StartsWith(pathSegment)))
                     {
-                        if (commonPath.Length == 0 && filePaths.All(str => str.ToString().StartsWith(pathSegment)))
-                        {
-                            commonPath = pathSegment;
-                        }
-                        else if (filePaths.All(str => str.ToString().StartsWith(commonPath + "/" + pathSegment)))
-                        {
-                            commonPath += "/" + pathSegment;
-                        }
-                        else
-                        {
-                            break;
-                        }
+                        commonPath = pathSegment;
                     }
-
-                    // Iterate all files and copy them.
-                    foreach (var filePath in filePaths)
+                    else if (filePaths.All(str => str.ToString().StartsWith(commonPath + "/" + pathSegment)))
                     {
-                        CopyFileCore(context, filePath, absoluteTargetDirectoryPath.GetFilePath(filePath), context.DirectoryExists(commonPath) ? commonPath : null);
+                        commonPath += "/" + pathSegment;
+                    }
+                    else
+                    {
+                        break;
                     }
                 }
-                else
+
+                // Iterate all files and copy them.
+                foreach (var filePath in filePaths)
                 {
-                    // Iterate all files and copy them.
-                    foreach (var filePath in filePaths)
-                    {
-                        CopyFileCore(context, filePath, absoluteTargetDirectoryPath.GetFilePath(filePath), null);
-                    }
+                    CopyFileCore(context, filePath, absoluteTargetDirectoryPath.GetFilePath(filePath), context.DirectoryExists(commonPath) ? commonPath : null);
+                }
+            }
+            else
+            {
+                // #1663: For empty enumerations, just return.
+                if (!filePaths.Any())
+                {
+                    return;
+                }
+
+                // Iterate all files and copy them.
+                foreach (var filePath in filePaths)
+                {
+                    CopyFileCore(context, filePath, absoluteTargetDirectoryPath.GetFilePath(filePath), null);
                 }
             }
         }

